packages/wallet/src/TxBuilder/types.ts

```diff
@@ -0,0 +1,126 @@
+// and that forces users to do many checks before being able to use any of it's properties and methods
+export type PartialTxOut = DeepPartial<Cardano.TxOut, Cardano.TokenMap>;
+
+export enum TxOutputFailure {
+  MinimumCoin = 'Minimum coin not met',
+  TokenBundleSizeExceedsLimit = 'Token Bundle Exceeds Limit',
+  MissingRequiredFields = 'Mandatory fields address or coin are missing'
+}
+
+export class OutputValidationMissingRequiredError extends CustomError {
+  public constructor(public txOut: PartialTxOut) {
+    super(TxOutputFailure.MissingRequiredFields);
+  }
+}
+
+export class OutputValidationMinimumCoinError extends CustomError {
+  public constructor(public txOut: PartialTxOut, public outputValidation: OutputValidation) {
+    super(TxOutputFailure.MinimumCoin);
+  }
+}
+
+export class OutputValidationTokenBundleSizeError extends CustomError {
+  public constructor(public txOut: PartialTxOut, public outputValidation: OutputValidation) {
+    super(TxOutputFailure.TokenBundleSizeExceedsLimit);
+  }
+}
+
+export class IncompatibleWalletError extends CustomError {}
+
+export type TxOutValidationError =
+  | OutputValidationMissingRequiredError
+  | OutputValidationMinimumCoinError
+  | OutputValidationTokenBundleSizeError;
+export type TxBodyValidationError = TxOutValidationError | InputSelectionError;
+
+export type Valid<TValid> = TValid & {
+  isValid: true;
+};
+
+export interface Invalid<TError> {
+  isValid: false;
+  errors: TError[];
+}
+
+export interface ValidTxOutData {
+  readonly txOut: Cardano.TxOut;
+}
+
+export type ValidTxOut = Valid<ValidTxOutData>;
+export type InvalidTxOut = Invalid<TxOutValidationError>;
+export type MaybeValidTxOut = ValidTxOut | InvalidTxOut;
+
+export interface OutputBuilder {
+  partialOutput: PartialTxOut;
+  value(value: Cardano.Value): OutputBuilder;
+  coin(coin: Cardano.Lovelace): OutputBuilder;
+  assets(assets: Cardano.TokenMap): OutputBuilder;
+  /**
+   * @param quantity To remove an asset, set quantity to 0
+   */
+  asset(assetId: Cardano.AssetId, quantity: bigint): OutputBuilder;
+  address(address: Cardano.Address): OutputBuilder;
+  datum(datum: Cardano.util.Hash32ByteBase16): OutputBuilder;
+  build(): Promise<MaybeValidTxOut>;
+}
+
+export interface SignedTx {
+  readonly tx: Cardano.NewTxAlonzo;
+  submit(): Promise<void>;
+}
+
+export interface ValidTxBody {
+  readonly body: Cardano.NewTxBodyAlonzo;
+  readonly auxiliaryData?: Cardano.AuxiliaryData;
+  readonly extraWitness?: Partial<Cardano.Witness>;
+
+  sign(props?: SignTransactionOptions): Promise<SignedTx>;
+}
+
+export type ValidTx = Valid<ValidTxBody>;
+export type InvalidTx = Invalid<TxBodyValidationError>;
+export type MaybeValidTx = ValidTx | InvalidTx;
+
+export interface TxBuilder {
+  partialTxBody: Partial<Cardano.NewTxBodyAlonzo>;
+  auxiliaryData?: Cardano.AuxiliaryData;
+  extraWitness?: Partial<Cardano.Witness>;
```

Comment:
This doesn't seem to be used anywhere. It should probably need to be changed into [these 2](https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/wallet/src/types.ts#L30-L31) props (same for `ValidTxBody` interface) that should then be passed to both `initializeTx` and `finalizeTx`

Reply:
[53ef574](https://github.com/input-output-hk/cardano-js-sdk/pull/396/commits/53ef574ab31271196145f491bc145e55c4774848) & [c4037b8](https://github.com/input-output-hk/cardano-js-sdk/pull/396/commits/c4037b828fc9dc3b4b1fad0446101aaa8f63f4cf)